src/services/compiler/host.ts

```diff
@@ -0,0 +1,136 @@
+
+interface RuntimeResources {
+    service: ChildProcess;
+    proxy: IPCProxy;
+}
+
+interface TestFunction {
+    (testRun: TestRun): Promise<unknown>;
+}
+
+export default class CompilerHost extends EventEmitter implements CompilerProtocol {
+    private runtime: Promise<RuntimeResources|undefined>;
+
+    public constructor () {
+        super();
+
+        this.runtime = Promise.resolve(void 0);
+    }
+
+    private _setupRoutes (proxy: IPCProxy): void {
+        proxy.register(this.executeAction, this);
+        proxy.register(this.ready, this);
+    }
+
+
+    private async _init (runtime: Promise<RuntimeResources|undefined>): Promise<RuntimeResources|undefined> {
+        const resolvedRuntime = await runtime;
+
+        if (resolvedRuntime)
+            return resolvedRuntime;
+
+        try {
+            const service = spawn(process.argv0, [SERVICE_PATH], { stdio: [0, 1, 2, 'pipe', 'pipe', 'pipe'] });
+
+            // HACK: Node.js definition are not correct when additional I/O channels are sp
+            const stdio = service.stdio as any;
+            const proxy = new IPCProxy(new HostTransport(stdio[HOST_INPUT_FD], stdio[HOST_OUTPUT_FD], stdio[HOST_SYNC_FD]));
+
+            this._setupRoutes(proxy);
+
+            await this.once('ready');
+
+            return { proxy, service };
+        }
+        catch (e) {
+            return void 0;
+        }
+    }
+
+    private async _getRuntime (): Promise<RuntimeResources> {
+        const runtime = await this.runtime;
+
+        if (!runtime)
+            throw new Error();
+
+        return runtime;
+    }
+
+    public async init (): Promise<void> {
+        this.runtime = this._init(this.runtime);
+
+        await this.runtime;
+    }
+
+
+    public async stop (): Promise<void> {
+        const { service } = await this._getRuntime();
+
+        service.kill();
+    }
+
+
+    private _wrapTestFunction (id: string, functionName: FunctionProperties): TestFunction {
+        return async testRun => {
+            try {
+                return await this.runTest({ id, functionName, testRunId: testRun.id });
+            }
+            catch (err) {
+                const errList = new TestCafeErrorList();
+
+                errList.addError(err);
+
+                throw errList;
+            }
+        };
+    }
+
+    public async ready (): Promise<void> {
+        this.emit('ready');
+    }
+
+    public async executeAction (data: ExecuteCommandArguments): Promise<unknown> {
+        if (!testRunTracker.activeTestRuns[data.id])
+            return void 0;
+
+        return testRunTracker
+            .activeTestRuns[data.id]
+            .executeAction(data.apiMethodName, data.command, data.callsite);
+    }
+
+    public async getTests ({ sourceList, compilerOptions }: CompilerArguments): Promise<Test[]> {
+        const { proxy } = await this._getRuntime();
+
+        const units = await proxy.call(this.getTests, { sourceList, compilerOptions });
```

Comment:
The first arg is function, so I supposed that this function will be called on callback.  However, inside the `proxy.call` method we need only function name. I would prefer to pass the `getTests` string or enum

Reply:
It's a quite low-level API. We should introduce [a higher level](https://github.com/DevExpress/testcafe/issues/4788) API and use some IDs instead of function names at this level. But I think we can do it as a separate task (#4788).